### PR TITLE
Issue#1318 delivery

### PR DESCRIPTION
**Pull request by Hugo-Hontans**
- [x] Use a CollectionService
- [x] Use CollectionGuards
- [x] Update akita ng fire

### CODE REVIEW

File: libs/material/src/lib/delivery/+state/delivery.model.ts [hunk 1]
```diff
@@ -41,3 +41,29 @@ export function createDelivery(params: Partial<Delivery>) {
     ...params
   } as Delivery;
 }
+
+export function timestampObjectsToDate(docs: any[]) {
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/delivery/+state/delivery.model.ts]: What about `fromTimestampToDate(docs: { date: Timestamp }[] = [])` ?
Like that you don't have to check if there is no docs

File: libs/material/src/lib/delivery/+state/delivery.model.ts [hunk 2]
```diff
@@ -41,3 +41,29 @@ export function createDelivery(params: Partial<Delivery>) {
...
+  if (!docs) {
+    return [];
+  }
+
+  return docs.map(doc => {
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/delivery/+state/delivery.model.ts]:
```typescript
return docs.map(doc => doc.date ? { ...doc, date: doc.date.toDate() } : doc);
```

File: libs/material/src/lib/delivery/+state/delivery.model.ts [hunk 3]
```diff
@@ -41,3 +41,29 @@ export function createDelivery(params: Partial<Delivery>) {
...
+    if (doc.date) {
+      return { ...doc, date: doc.date.toDate() };
+    } else {
+      return doc;
+    }
+  });
+}
+
+/** Takes a DeliveryDB (dates in Timestamp) and returns a Delivery with dates in type Date */
+export function convertDeliveryWithTimestampsToDelivery(delivery: DeliveryWithTimestamps): Delivery {
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/delivery/+state/delivery.model.ts]: Call it : `createDeliveryFromFirestore` to make it easy to read, and coherent with akita-ng-fire

File: libs/material/src/lib/delivery/+state/delivery.service.ts [hunk 4]
```diff
@@ -26,36 +26,34 @@ interface AddDeliveryOptions {
...
-  const mgDeadlines = delivery.mgDeadlines || [];
-
-  return {
-    ...delivery,
-    dueDate: delivery.dueDate ? delivery.dueDate.toDate() : null,
-    steps: timestampObjectsToDate(delivery.steps),
-    mgDeadlines: timestampObjectsToDate(mgDeadlines)
-  };
-}
+export const deliveryQuery = (deliveryId: string): Query<DeliveryWithTimestamps> => ({
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/delivery/+state/delivery.service.ts]: There is two const with almost the same name. This is difficult to understand.

File: libs/material/src/lib/delivery/+state/delivery.service.ts [hunk 5]
```diff
@@ -228,91 +229,56 @@ export class DeliveryService {
...
       oldStep => !stepsWithId.some(newStep => newStep.id === oldStep.id)
     );
 
-    // Remove stepId from the materials according to this array
-    this.removeMaterialsStepId(deletedSteps, batch);
-
-    return batch.update(deliveryRef, { steps: stepsWithId });
-  }
+    // We set the concerned materials stepId to an empty string
+    deletedSteps.forEach(step => {
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/delivery/+state/delivery.service.ts]: I have a hard time to understand what you're doing here. But I don't really like the idea of removing all those documents without using a batch....

File: libs/material/src/lib/delivery/+state/delivery.service.ts [hunk 6]
```diff
@@ -228,91 +229,56 @@ export class DeliveryService {
...
+        const materialsWithoutStep = materials.map(material => ({ ...material, stepId: ''}));
+        this.materialService.updateMovieMaterials(materialsWithoutStep, this.movieQuery.getActiveId());
+      }
+    })
 
-  /** Remove stepId of materials of delivery for an array of steps */
-  private removeMaterialsStepId(steps: Step[], batch: firebase.firestore.WriteBatch) {
-    // TODO(issue#773): Use a transaction to make sure we don't lose data
-    const deliveryId = this.query.getActiveId();
+    return this.update(this.query.getActiveId(), {
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/delivery/+state/delivery.service.ts]: You're repeating `this.query.getActive` and `...getActiveId` several time. Create a const `delivery` once and for all in the function and use it everywhere

File: libs/material/src/lib/delivery/+state/delivery.store.ts [hunk 7]
```diff
@@ -27,7 +28,7 @@ export interface DeliveryWizard {
   options: DeliveryOption[];
 }
 
-export interface DeliveryState extends EntityState<Delivery>, ActiveState<string> {
+export interface DeliveryState extends CollectionState<Delivery> {
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/delivery/+state/delivery.store.ts]: Don't use CollectionState use `EntityState` and `ActivateState` instead.

File: libs/material/src/lib/delivery/+state/delivery.store.ts [hunk 8]
```diff
@@ -44,6 +45,16 @@ export class DeliveryStore extends EntityStore<DeliveryState, Delivery> {
     super(initialState);
   }
 
+  // TODO: #issue1288, type correctly delivery
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/delivery/+state/delivery.store.ts]: I think, this has been solve in the latest version of akita

File: libs/material/src/lib/material/+state/material.service.ts [hunk 9]
```diff
@@ -135,10 +135,16 @@ export class MaterialService extends CollectionService<MaterialState> {
     materials.forEach(material => this.update(material.id, { isPaid: !material.isPaid }));
   }
 
-  /** Update materials of a movie (specific fields like 'owner' and 'storage'). */
+  /** Update materials of a movie (specific fields like 'owner', 'storage', 'stepId'). */
   public updateMovieMaterials(materials: Material[], movieId: string) {
     this.subcollectionPath = `movies/${movieId}/materials`;
-      materials.forEach(material => this.update(material.id, material));
+    materials.forEach(material => this.update(material.id, material));
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/material/+state/material.service.ts]: I don't like the idea of doing that many update without a batch. Use `this.update(materials)` directly

File: libs/material/src/lib/material/+state/material.service.ts [hunk 10]
```diff
@@ -135,10 +135,16 @@ export class MaterialService extends CollectionService<MaterialState> {
...
+  }
+
+  /** Update stepId of materials of a delivery to empty string. */
+  public removeStepIdDeliveryMaterials(materials: Material[], deliveryId: string) {
+    this.subcollectionPath = `deliveries/${deliveryId}/materials`;
+    materials.forEach(material => this.update(material.id, { stepId: '' }));
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/material/+state/material.service.ts]: Same here

File: libs/material/src/lib/material/+state/material.service.ts [hunk 11]
```diff
@@ -242,6 +248,18 @@ export class MaterialService extends CollectionService<MaterialState> {
     return this.getValue();
   }
 
+  /** Returns a snapshot of movie materials and sets the subcollectionPath. */
+  public async getMovieMaterials(movieId: string) {
+    this.subcollectionPath = `movies/${movieId}/materials`;
```
Comment (GrandSchtroumpf) [inline libs/material/src/lib/material/+state/material.service.ts]: What is subcollectionpath? What if I call getValue from somewhere else just after ? This seems super dangerous.

**Comment by Hugo-Hontans** [comment]
Replaced by PR #1357

**Pull request closed without merging**